hngprojects/hng_boilerplate_nestjs
# feat: creating a blog by superadmin

## What does the PR do?
This PR introduces a new endpoint for creating blog posts that is restricted to super admins. The endpoint only allows access if the provided token is valid and has the super admin role. The PR ensures that:

- [x] The endpoint handles missing or invalid authorization headers with a 401 status code
- [x] It returns a 403 status code if the token does not have the super admin role.
- [x] It correctly validates input fields and responds with a 400 status code for invalid requests.
- [x] On successful blog creation, it returns a 201 status code and the appropriate response body.


## How should this be manually tested?
1. Ensure you have a valid super admin token.
2. Send a POST request to the following endpoint:
POST /api/v1/blogs

Include the Authorization header:
 ```json

 Authorization: Bearer <access_token>
 ```
3. Provide a valid request body:
```
{
  "title": "Blog Title",
  "content": "<p>First paragraph.</p><p>Second paragraph.</p>",
  "tags": ["tag1", "tag2"],
  "image_urls": ["https://example.com/image1.jpg", "https://example.com/image2.jpg"]
}
```

### Success Response
on a successful execution, you will receive the following response structure:
![blog_success](https://github.com/user-attachments/assets/b264fd07-c160-4e68-a0a6-812e6e3bff4a)

```
{
    {
    "status_code": 201,
    "message": "Blog created successfully",
    "data": {
        "blog_id": "unique-blog-id",
        "title": "Blog Title",
        "content": "<p>First paragraph.</p><p>Second paragraph.</p>",
        "image_urls": ["https://example.com/image1.jpg", "https://example.com/image2.jpg"],
        "tags": ["tag1", "tag2"],
        "author": "Super Admin",
        "created_at": "2024-08-05T19:03:52.194Z"
    }
}
}
```
### "Invalid Input" Response
If the request is invalid, you will receive the following response:
![blog_invalidinput](https://github.com/user-attachments/assets/7c9cd82f-e9c9-4014-b8c6-f8a8003b749e)

```
{
    "status_code": 400,
    "message": "Invalid request. Title and content are required."
}


```
### Unauthorized Response
![blog_unauthorized](https://github.com/user-attachments/assets/ec84af38-b8fa-434e-8b75-44dd201c2687)

```
{
    "status_code": 401,
    "message": "Unauthorized. Please log in."
}
```

### "Internal Server Error" Response

```
{
    "status_code": 500,
    "message": "Internal server error. Failed to create the blog post."
}
```

## Task
- [x]  Implemented POST /api/v1/blogs endpoint in blog.controller.ts.
- [x] Added method in BlogService to handle blog creation with super admin restriction.
- [x] Validated token and user role to ensure only super admins can create blogs.
- [x] Implemented appropriate error handling for invalid tokens and request bodies.. 
- [x] Added unit tests for the new endpoint covering authorization, input validation, and success cases
- [x] Updated API documentation to reflect the new endpoint and its requirements.

##Unit Test Evidence
![blog_test](https://github.com/user-attachments/assets/e577aa50-87b1-4224-9c46-128d7d9c647b)

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-126/endpoint-for-creating-a-blog-or-46)
